File: lab3/zad1/cpp/queue.cpp
#include <iostream>
#include <fstream>
#include "node.h"
#include "queue.h"

using namespace std;

Queue::Queue() {
    front = nullptr;
    rear = nullptr;
}

Queue::~Queue() {
    while (front != nullptr) {
        QueueNode* temp = front;
        front = front->next;
        delete temp;
    }
}

void Queue::enqueue(const string& val) {
    auto *newNode = new QueueNode(val);

    if (front == nullptr) {
        front = newNode;
        rear = newNode;
    } else {
        rear->next = newNode;
        rear = newNode;
    }
}

void Queue::dequeue() {
    if (front == nullptr) {
        throw runtime_error("error");
    }
    QueueNode* temp = front;
    front = front->next;
    
    if (front == nullptr) {
        rear = nullptr;
    }
    delete temp;
}

void Queue::print() const {
    if (front == nullptr) {
        cout << "error" << '\n';
        return;
    }
    QueueNode* curr = front;
    cout << "очередь: ";
    while (curr != nullptr) {
        cout << curr->data;
        if (curr->next != nullptr) {
            cout << " -> ";
        }
        curr = curr->next;
    }
    cout << '\n';
}

void Queue::saveToFile(ofstream& file) const {
    if (!file.is_open()) {
        return;
    }
    QueueNode* curr = front;
    while (curr != nullptr) {
        file << curr->data << " ";
        curr = curr->next;
    }
}

void Queue::loadFromFile(ifstream& file) {
    if (!file.is_open()) {
        return;
    }
    string value;
    while (file >> value) {
        enqueue(value);
    }
}

void Queue::saveToBinaryFile(ofstream& file) const {
    if (!file.is_open()) {
        return;
    }
    QueueNode* curr = front;
    while (curr != nullptr) {
        size_t strSize = curr->data.size();
        auto streamSize = static_cast<streamsize>(strSize);
        file.write(reinterpret_cast<const char*>(&strSize), sizeof(strSize));
        file.write(curr->data.c_str(), streamSize);
        curr = curr->next;
    }
}

void Queue::loadFromBinaryFile(ifstream& file) {
    if (!file.is_open()) {
        return;
    }
    while (file.peek() != EOF) {
        size_t strSize;
        file.read(reinterpret_cast<char*>(&strSize), sizeof(strSize));
        if (file.eof()){ 
            break;
        }
        auto streamSize = static_cast<streamsize>(strSize);
        char* buffer = new char[strSize + 1];
        file.read(buffer, streamSize);
        buffer[strSize] = '\0';
        enqueue(string(buffer));
        delete[] buffer;
    }
}

File: lab3/zad2/go/queue/bench/queue.go
package bench

import (
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"os"
	"strings"
)

type Queue struct {
	front *QueueNode
	rear  *QueueNode
}

func NewQueue() *Queue {
	return &Queue{}
}

func (q *Queue) Enqueue(val string) {
	newNode := &QueueNode{Data: val}

	if q.front == nil {
		q.front = newNode
		q.rear = newNode
	} else {
		q.rear.Next = newNode
		q.rear = newNode
	}
}

func (q *Queue) Dequeue() error {
	if q.front == nil {
		return errors.New("error")
	}

	q.front = q.front.Next
	if q.front == nil {
		q.rear = nil
	}
	return nil
}

func (q *Queue) Print() {
	if q.front == nil {
		fmt.Println("error")
		return
	}

	curr := q.front
	fmt.Print("очередь: ")
	for curr != nil {
		fmt.Print(curr.Data)
		if curr.Next != nil {
			fmt.Print(" -> ")
		}
		curr = curr.Next
	}
	fmt.Println()
}

func (q *Queue) SaveToFile(file *os.File) error {
	if file == nil {
		return errors.New("file is nil")
	}

	curr := q.front
	for curr != nil {
		_, err := file.WriteString(curr.Data + " ")
		if err != nil {
			return err
		}
		curr = curr.Next
	}
	return nil
}

func (q *Queue) LoadFromFile(file *os.File) error {
	if file == nil {
		return errors.New("file is nil")
	}

	content, err := io.ReadAll(file)
	if err != nil {
		return err
	}

	values := strings.Fields(string(content))
	for _, val := range values {
		q.Enqueue(val)
	}
	return nil
}

func (q *Queue) SaveToBinaryFile(file *os.File) error {
	if file == nil {
		return errors.New("file is nil")
	}

	curr := q.front
	for curr != nil {
		strSize := uint64(len(curr.Data))
		err := binary.Write(file, binary.LittleEndian, strSize)
		if err != nil {
			return err
		}
		_, err = file.WriteString(curr.Data)
		if err != nil {
			return err
		}
		curr = curr.Next
	}
	return nil
}

func (q *Queue) LoadFromBinaryFile(file *os.File) error {
	if file == nil {
		return errors.New("file is nil")
	}

	for {
		var strSize uint64
		err := binary.Read(file, binary.LittleEndian, &strSize)
		if err != nil {
			if err == io.EOF {
				break
			}
			return err
		}

		buffer := make([]byte, strSize)
		_, err = file.Read(buffer)
		if err != nil {
			return err
		}
		q.Enqueue(string(buffer))
	}
	return nil
}
